datafusion/optimizer/src/unwrap_cast_in_comparison.rs

```diff
@@ -140,140 +139,132 @@ struct UnwrapCastExprRewriter {
+                        return Ok(Transformed::no(expr));
                     };
+                    let Ok(right_type) = right.get_type(&self.schema) else {
+                        return Ok(Transformed::no(expr));
+                    };
+                    is_support_data_type(&left_type)
+                        && is_support_data_type(&right_type)
+                        && is_comparison_op(op)
+                } =>
+            {
+                match (left.as_mut(), right.as_mut()) {
+                    (
+                        Expr::Literal(left_lit_value),
+                        Expr::TryCast(TryCast {
+                            expr: right_expr, ..
+                        })
+                        | Expr::Cast(Cast {
+                            expr: right_expr, ..
+                        }),
+                    ) => {
+                        // if the left_lit_value can be casted to the type of expr
+                        // we need to unwrap the cast for cast/try_cast expr, and add cast to the literal
+                        let Ok(expr_type) = right_expr.get_type(&self.schema) else {
+                            return Ok(Transformed::no(expr));
+                        };
+                        let Ok(Some(value)) =
+                            try_cast_literal_to_type(left_lit_value, &expr_type)
+                        else {
+                            return Ok(Transformed::no(expr));
+                        };
+                        **left = lit(value);
+                        // unwrap the cast/try_cast for the right expr
+                        **right =
+                            mem::replace(right_expr, Expr::Literal(ScalarValue::Null));
+                        Ok(Transformed::yes(expr))
+                    }
+                    (
+                        Expr::TryCast(TryCast {
+                            expr: left_expr, ..
+                        })
+                        | Expr::Cast(Cast {
+                            expr: left_expr, ..
+                        }),
+                        Expr::Literal(right_lit_value),
+                    ) => {
+                        // if the right_lit_value can be casted to the type of expr
+                        // we need to unwrap the cast for cast/try_cast expr, and add cast to the literal
+                        let Ok(expr_type) = left_expr.get_type(&self.schema) else {
+                            return Ok(Transformed::no(expr));
+                        };
+                        let Ok(Some(value)) =
+                            try_cast_literal_to_type(right_lit_value, &expr_type)
+                        else {
+                            return Ok(Transformed::no(expr));
+                        };
+                        // unwrap the cast/try_cast for the left expr
+                        **left =
+                            mem::replace(left_expr, Expr::Literal(ScalarValue::Null));
```

Comment:
Would be nice to have a default `Expr` so that we could use `mem::take()`.

Reply:
Oh yeah!

Reply:
Shall I open a follow-up PR that adds default for `Expr`?

Reply:
`mem::take` is a nice thing to avoid cloning, I think I used it to avoid String cloning other day, and we can have a default Expr with panicking inside I'd say...

Reply:
I've opened a minor PR to add the default `Expr`: https://github.com/apache/arrow-datafusion/pull/10127, but as `Expr` is just an enum so I don't know how to panic inside it.